File: src/tools/notion.ts
import type { PageObjectResponse } from "@notionhq/client/build/src/api-endpoints.js";
import { isFullPage } from "@notionhq/client";
import { config } from "../config.js";
import { getNotion } from "../integrations/notion.js";
import type { ToolDef } from "../agent/tool-types.js";
import { PROJECT_PROPS, extractProjectBriefFields } from "../lib/notion-project-fields.js";

/** Project names match `PROJECT_PROPS` in `src/lib/notion-project-fields.ts`. */
const P = {
  projectTitle: PROJECT_PROPS.title,
  projectStatus: PROJECT_PROPS.status,
  projectDeadline: PROJECT_PROPS.deadline,
  projectPriority: PROJECT_PROPS.priority,
  projectCurrentFocus: PROJECT_PROPS.currentFocus,
  projectNextAction: PROJECT_PROPS.nextAction,
  taskTitle: "Name",
  taskDue: "Due",
  taskStatus: "Status",
  taskProject: "Project",
} as const;

function requireProjectsDb() {
  if (!config.NOTION_PROJECTS_DB_ID) throw new Error("NOTION_PROJECTS_DB_ID not configured");
}

function requireTasksDb() {
  if (!config.NOTION_TASKS_DB_ID) throw new Error("NOTION_TASKS_DB_ID not configured");
}

function titleFromPage(page: PageObjectResponse): string {
  for (const [, prop] of Object.entries(page.properties)) {
    if (
      prop &&
      typeof prop === "object" &&
      "type" in prop &&
      prop.type === "title" &&
      "title" in prop &&
      Array.isArray((prop as { title: { plain_text: string }[] }).title)
    ) {
      const title = (prop as { title: { plain_text: string }[] }).title;
      return title.map((t: { plain_text: string }) => t.plain_text).join("") || "Untitled";
    }
  }
  return "Untitled";
}

function priorityPayload(value: string): { select: { name: string } } | { number: number } {
  const n = Number(value);
  if (value.trim() !== "" && !Number.isNaN(n) && String(n) === value.trim()) {
    return { number: n };
  }
  return { select: { name: value } };
}

function mapProjectRow(page: PageObjectResponse) {
  const brief = extractProjectBriefFields(page);
  return {
    id: page.id,
    ...brief,
    url: page.url,
  };
}

function mapTaskRow(page: PageObjectResponse) {
  const props = page.properties;
  const name = titleFromPage(page);
  let status = "";
  const st = props[P.taskStatus];
  if (st?.type === "status") status = st.status?.name ?? "";
  let due: string | null = null;
  const d = props[P.taskDue];
  if (d?.type === "date") due = d.date?.start ?? null;
  let projectId: string | null = null;
  const rel = props[P.taskProject];
  if (rel?.type === "relation" && rel.relation[0]) projectId = rel.relation[0].id;
  return { id: page.id, name, status, due, projectPageId: projectId, url: page.url };
}

function mapPages(results: unknown[]) {
  return results
    .filter((r): r is PageObjectResponse => isFullPage(r as Parameters<typeof isFullPage>[0]))
    .map((p) => mapProjectRow(p));
}

function mapTaskPages(results: unknown[]) {
  return results
    .filter((r): r is PageObjectResponse => isFullPage(r as Parameters<typeof isFullPage>[0]))
    .map((p) => mapTaskRow(p));
}

export const notionTools: Record<string, ToolDef> = {
  notion_search_projects: {
    spec: {
      type: "function",
      function: {
        name: "notion_search_projects",
        description:
          "Search projects: returns name, status, priority, current focus, next action, deadline, id. Optional status filter.",
        parameters: {
          type: "object",
          properties: {
            status: { type: "string", description: "Optional status name to filter (exact)" },
            limit: { type: "integer", description: "Max results (default 20)" },
          },
        },
      },
    },
    handler: async (args: { status?: string; limit?: number }) => {
      requireProjectsDb();
      const notion = getNotion();
      const filter = args.status
        ? { property: P.projectStatus, status: { equals: args.status } }
        : undefined;
      const res = await notion.databases.query({
        database_id: config.NOTION_PROJECTS_DB_ID,
        filter: filter as never,
        page_size: Math.min(args.limit ?? 20, 100),
      });
      return mapPages(res.results);
    },
  },

  notion_create_project: {
    spec: {
      type: "function",
      function: {
        name: "notion_create_project",
        description: "Create a new project page in the Projects database.",
        parameters: {
          type: "object",
          properties: {
            name: { type: "string", description: "Project title" },
            status: { type: "string", description: "Status option name (must exist in Notion)" },
            deadline: { type: "string", description: "ISO date YYYY-MM-DD or empty" },
            priority: { type: "string", description: "Select name or numeric string if Priority is a number" },
            current_focus: { type: "string", description: "Current focus (rich text)" },
            next_action: { type: "string", description: "Next action (rich text)" },
          },
          required: ["name"],
        },
      },
    },
    handler: async (args: {
      name: string;
      status?: string;
      deadline?: string;
      priority?: string;
      current_focus?: string;
      next_action?: string;
    }) => {
      requireProjectsDb();
      const notion = getNotion();
      const props: Record<string, unknown> = {
        [P.projectTitle]: {
          title: [{ text: { content: args.name } }],
        },
      };
      if (args.status) {
        props[P.projectStatus] = { status: { name: args.status } };
      }
      if (args.deadline) {
        props[P.projectDeadline] = { date: { start: args.deadline } };
      }
      if (args.priority) {
        props[P.projectPriority] = priorityPayload(args.priority) as never;
      }
      if (args.current_focus) {
        props[P.projectCurrentFocus] = {
          rich_text: [{ text: { content: args.current_focus } }],
        };
      }
      if (args.next_action) {
        props[P.projectNextAction] = {
          rich_text: [{ text: { content: args.next_action } }],
        };
      }
      const created = await notion.pages.create({
        parent: { database_id: config.NOTION_PROJECTS_DB_ID },
        properties: props as never,
      });
      if (!isFullPage(created)) return { id: created.id, url: (created as { url?: string }).url };
      return mapProjectRow(created);
    },
  },

  notion_update_project: {
    spec: {
      type: "function",
      function: {
        name: "notion_update_project",
        description: "Update an existing project Notion page by page ID.",
        parameters: {
          type: "object",
          properties: {
            page_id: { type: "string", description: "Notion page UUID" },
            name: { type: "string" },
            status: { type: "string" },
            deadline: { type: "string", description: "YYYY-MM-DD or empty to clear" },
            priority: { type: "string", description: "Set/clear priority (empty string clears select; number as string for number columns)" },
            current_focus: { type: "string" },
            next_action: { type: "string" },
          },
          required: ["page_id"],
        },
      },
    },
    handler: async (args: {
      page_id: string;
      name?: string;
      status?: string;
      deadline?: string;
      priority?: string;
      current_focus?: string;
      next_action?: string;
    }) => {
      requireProjectsDb();
      const notion = getNotion();
      const props: Record<string, unknown> = {};
      if (args.name !== undefined) {
        props[P.projectTitle] = { title: [{ text: { content: args.name } }] };
      }
      if (args.status !== undefined) {
        props[P.projectStatus] = { status: { name: args.status } };
      }
      if (args.deadline !== undefined) {
        props[P.projectDeadline] = args.deadline
          ? { date: { start: args.deadline } }
          : { date: null };
      }
      if (args.priority !== undefined) {
        props[P.projectPriority] = (
          args.priority === ""
            ? { select: null }
            : priorityPayload(args.priority)
        ) as never;
      }
      if (args.current_focus !== undefined) {
        props[P.projectCurrentFocus] =
          args.current_focus === ""
            ? { rich_text: [] }
            : { rich_text: [{ text: { content: args.current_focus } }] };
      }
      if (args.next_action !== undefined) {
        props[P.projectNextAction] =
          args.next_action === ""
            ? { rich_text: [] }
            : { rich_text: [{ text: { content: args.next_action } }] };
      }
      const updated = await notion.pages.update({
        page_id: args.page_id,
        properties: props as never,
      });
      if (!isFullPage(updated)) return { id: updated.id, ok: true };
      return mapProjectRow(updated);
    },
  },

  notion_search_tasks: {
    spec: {
      type: "function",
      function: {
        name: "notion_search_tasks",
        description: "Search tasks in the Tasks Notion database.",
        parameters: {
          type: "object",
          properties: {
            status: { type: "string", description: "Optional status filter" },
            limit: { type: "integer" },
          },
        },
      },
    },
    handler: async (args: { status?: string; limit?: number }) => {
      requireTasksDb();
      const notion = getNotion();
      const filter = args.status
        ? { property: P.taskStatus, status: { equals: args.status } }
        : undefined;
      const res = await notion.databases.query({
        database_id: config.NOTION_TASKS_DB_ID,
        filter: filter as never,
        page_size: Math.min(args.limit ?? 20, 100),
      });
      return mapTaskPages(res.results);
    },
  },

  notion_create_task: {
    spec: {
      type: "function",
      function: {
        name: "notion_create_task",
        description: "Create a task in the Tasks database. Optionally link to a project page ID.",
        parameters: {
          type: "object",
          properties: {
            title: { type: "string" },
            due: { type: "string", description: "YYYY-MM-DD" },
            status: { type: "string" },
            project_page_id: { type: "string", description: "Parent project Notion page UUID" },
          },
          required: ["title"],
        },
      },
    },
    handler: async (args: {
      title: string;
      due?: string;
      status?: string;
      project_page_id?: string;
    }) => {
      requireTasksDb();
      const notion = getNotion();
      const props: Record<string, unknown> = {
        [P.taskTitle]: {
          title: [{ text: { content: args.title } }],
        },
      };
      if (args.due) props[P.taskDue] = { date: { start: args.due } };
      if (args.status) props[P.taskStatus] = { status: { name: args.status } };
      if (args.project_page_id) {
        props[P.taskProject] = { relation: [{ id: args.project_page_id }] };
      }
      const created = await notion.pages.create({
        parent: { database_id: config.NOTION_TASKS_DB_ID },
        properties: props as never,
      });
      if (!isFullPage(created)) return { id: created.id };
      return mapTaskRow(created);
    },
  },

  notion_update_task: {
    spec: {
      type: "function",
      function: {
        name: "notion_update_task",
        description: "Update a task page by ID.",
        parameters: {
          type: "object",
          properties: {
            page_id: { type: "string" },
            title: { type: "string" },
            due: { type: "string" },
            status: { type: "string" },
            project_page_id: { type: "string", description: "Set or replace project relation" },
          },
          required: ["page_id"],
        },
      },
    },
    handler: async (args: {
      page_id: string;
      title?: string;
      due?: string;
      status?: string;
      project_page_id?: string;
    }) => {
      requireTasksDb();
      const notion = getNotion();
      const props: Record<string, unknown> = {};
      if (args.title !== undefined) {
        props[P.taskTitle] = { title: [{ text: { content: args.title } }] };
      }
      if (args.due !== undefined) {
        props[P.taskDue] = args.due ? { date: { start: args.due } } : { date: null };
      }
      if (args.status !== undefined) {
        props[P.taskStatus] = { status: { name: args.status } };
      }
      if (args.project_page_id !== undefined) {
        props[P.taskProject] = { relation: [{ id: args.project_page_id }] };
      }
      const updated = await notion.pages.update({
        page_id: args.page_id,
        properties: props as never,
      });
      if (!isFullPage(updated)) return { id: updated.id, ok: true };
      return mapTaskRow(updated);
    },
  },
};

File: src/tools/notion.py
from typing import Any

from src.config import get_settings
from src.exceptions import ConfigError
from src.integrations.notion import get_notion
from src.lib.notion_project_fields import (
    PROJECT_PROPS,
    extract_project_brief_fields,
    is_full_page,
    notion_status_clear_payload,
    notion_status_db_filter,
    notion_status_update_payload,
)
from src.lib.notion_task_fields import TASK_PROPS, map_task_row

P = {
    "projectTitle": PROJECT_PROPS["title"],
    "projectStatus": PROJECT_PROPS["status"],
    "projectDeadline": PROJECT_PROPS["deadline"],
    "projectPriority": PROJECT_PROPS["priority"],
    "projectCurrentFocus": PROJECT_PROPS["currentFocus"],
    "projectNextAction": PROJECT_PROPS["nextAction"],
    "taskTitle": TASK_PROPS["title"],
    "taskDue": TASK_PROPS["due"],
    "taskStatus": TASK_PROPS["status"],
    "taskProject": TASK_PROPS["project"],
}


def _require_projects_db() -> str:
    db_id = get_settings().NOTION_PROJECTS_DB_ID
    if not db_id:
        raise ConfigError("NOTION_PROJECTS_DB_ID not configured")
    return db_id


def _require_tasks_db() -> str:
    db_id = get_settings().NOTION_TASKS_DB_ID
    if not db_id:
        raise ConfigError("NOTION_TASKS_DB_ID not configured")
    return db_id


def _priority_payload(value: str) -> dict[str, Any]:
    stripped = value.strip()
    try:
        number = float(stripped)
    except ValueError:
        return {"select": {"name": value}}
    if stripped and str(int(number) if number.is_integer() else number) == stripped:
        return {"number": int(number) if number.is_integer() else number}
    return {"select": {"name": value}}


def _map_project_row(page: dict[str, Any]) -> dict[str, Any]:
    return {"id": page.get("id"), **extract_project_brief_fields(page), "url": page.get("url")}


def _map_pages(results: list[Any]) -> list[dict[str, Any]]:
    return [_map_project_row(r) for r in results if is_full_page(r)]


def _map_task_pages(results: list[Any]) -> list[dict[str, Any]]:
    return [map_task_row(r) for r in results if is_full_page(r)]


async def _search_projects(args: dict[str, Any]) -> list[dict[str, Any]]:
    db_id = _require_projects_db()
    notion = get_notion()
    status = args.get("status")
    filt = notion_status_db_filter(P["projectStatus"], status) if status else None
    res = await notion.databases.query(
        database_id=db_id,
        **({"filter": filt} if filt else {}),
        page_size=min(args.get("limit") or 20, 100),
    )
    return _map_pages(res.get("results") or [])


async def _create_project(args: dict[str, Any]) -> dict[str, Any]:
    db_id = _require_projects_db()
    notion = get_notion()
    props: dict[str, Any] = {
        P["projectTitle"]: {"title": [{"text": {"content": args["name"]}}]},
    }
    if args.get("status"):
        props[P["projectStatus"]] = notion_status_update_payload(args["status"])
    if args.get("deadline"):
        props[P["projectDeadline"]] = {"date": {"start": args["deadline"]}}
    if args.get("priority"):
        props[P["projectPriority"]] = _priority_payload(args["priority"])
    if args.get("current_focus"):
        props[P["projectCurrentFocus"]] = {
            "rich_text": [{"text": {"content": args["current_focus"]}}]
        }
    if args.get("next_action"):
        props[P["projectNextAction"]] = {"rich_text": [{"text": {"content": args["next_action"]}}]}
    created = await notion.pages.create(parent={"database_id": db_id}, properties=props)
    if not is_full_page(created):
        return {"id": created.get("id"), "url": created.get("url")}
    return _map_project_row(created)


async def _update_project(args: dict[str, Any]) -> dict[str, Any]:
    _require_projects_db()
    notion = get_notion()
    props: dict[str, Any] = {}
    if args.get("name") is not None:
        props[P["projectTitle"]] = {"title": [{"text": {"content": args["name"]}}]}
    if "status" in args:
        props[P["projectStatus"]] = (
            notion_status_clear_payload()
            if args["status"] == ""
            else notion_status_update_payload(args["status"])
        )
    if "deadline" in args:
        props[P["projectDeadline"]] = (
            {"date": {"start": args["deadline"]}} if args["deadline"] else {"date": None}
        )
    if "priority" in args:
        props[P["projectPriority"]] = (
            {"select": None} if args["priority"] == "" else _priority_payload(args["priority"])
        )
    if "current_focus" in args:
        props[P["projectCurrentFocus"]] = (
            {"rich_text": []}
            if args["current_focus"] == ""
            else {"rich_text": [{"text": {"content": args["current_focus"]}}]}
        )
    if "next_action" in args:
        props[P["projectNextAction"]] = (
            {"rich_text": []}
            if args["next_action"] == ""
            else {"rich_text": [{"text": {"content": args["next_action"]}}]}
        )
    updated = await notion.pages.update(page_id=args["page_id"], properties=props)
    if not is_full_page(updated):
        return {"id": updated.get("id"), "ok": True}
    return _map_project_row(updated)


async def _search_tasks(args: dict[str, Any]) -> list[dict[str, Any]]:
    db_id = _require_tasks_db()
    notion = get_notion()
    status = args.get("status")
    filt = notion_status_db_filter(P["taskStatus"], status) if status else None
    res = await notion.databases.query(
        database_id=db_id,
        **({"filter": filt} if filt else {}),
        page_size=min(args.get("limit") or 20, 100),
    )
    return _map_task_pages(res.get("results") or [])


async def _create_task(args: dict[str, Any]) -> dict[str, Any]:
    db_id = _require_tasks_db()
    notion = get_notion()
    props: dict[str, Any] = {
        P["taskTitle"]: {"title": [{"text": {"content": args["title"]}}]},
    }
    if args.get("due"):
        props[P["taskDue"]] = {"date": {"start": args["due"]}}
    if args.get("status"):
        props[P["taskStatus"]] = notion_status_update_payload(args["status"])
    if args.get("project_page_id"):
        props[P["taskProject"]] = {"relation": [{"id": args["project_page_id"]}]}
    created = await notion.pages.create(parent={"database_id": db_id}, properties=props)
    if not is_full_page(created):
        return {"id": created.get("id")}
    return map_task_row(created)


async def _update_task(args: dict[str, Any]) -> dict[str, Any]:
    _require_tasks_db()
    notion = get_notion()
    props: dict[str, Any] = {}
    if "title" in args:
        props[P["taskTitle"]] = {"title": [{"text": {"content": args["title"]}}]}
    if "due" in args:
        props[P["taskDue"]] = {"date": {"start": args["due"]}} if args["due"] else {"date": None}
    if "status" in args:
        props[P["taskStatus"]] = (
            notion_status_clear_payload()
            if args["status"] == ""
            else notion_status_update_payload(args["status"])
        )
    if "project_page_id" in args:
        props[P["taskProject"]] = {"relation": [{"id": args["project_page_id"]}]}
    updated = await notion.pages.update(page_id=args["page_id"], properties=props)
    if not is_full_page(updated):
        return {"id": updated.get("id"), "ok": True}
    return map_task_row(updated)


async def _describe_tasks_database(_args: dict[str, Any]) -> dict[str, Any]:
    db_id = _require_tasks_db()
    notion = get_notion()
    db = await notion.databases.retrieve(database_id=db_id)
    properties: dict[str, dict[str, Any]] = {}
    for name, prop in (db.get("properties") or {}).items():
        kind = prop.get("type")
        if kind == "select":
            properties[name] = {
                "type": "select",
                "options": [o.get("name") for o in (prop.get("select") or {}).get("options") or []],
            }
        elif kind == "status":
            properties[name] = {
                "type": "status",
                "options": [o.get("name") for o in (prop.get("status") or {}).get("options") or []],
            }
        elif kind == "relation":
            properties[name] = {
                "type": "relation",
                "database_id": (prop.get("relation") or {}).get("database_id"),
            }
        else:
            properties[name] = {"type": kind}
    return {"tasks_database_id": db_id, "properties": properties}


notion_tools = {
    "notion_search_projects": {
        "spec": {
            "type": "function",
            "function": {
                "name": "notion_search_projects",
                "description": (
                    "Search projects: returns name, status, priority, current focus, "
                    "next action, deadline, id. Optional status filter."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "status": {
                            "type": "string",
                            "description": "Optional status name to filter (exact)",
                        },
                        "limit": {"type": "integer", "description": "Max results (default 20)"},
                    },
                },
            },
        },
        "handler": _search_projects,
    },
    "notion_create_project": {
        "spec": {
            "type": "function",
            "function": {
                "name": "notion_create_project",
                "description": "Create a new project page in the Projects database.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string", "description": "Project title"},
                        "status": {
                            "type": "string",
                            "description": "Status option name (must exist in Notion)",
                        },
                        "deadline": {
                            "type": "string",
                            "description": "ISO date YYYY-MM-DD or empty",
                        },
                        "priority": {
                            "type": "string",
                            "description": "Select name or numeric string if Priority is a number",
                        },
                        "current_focus": {
                            "type": "string",
                            "description": "Current focus (rich text)",
                        },
                        "next_action": {"type": "string", "description": "Next action (rich text)"},
                    },
                    "required": ["name"],
                },
            },
        },
        "handler": _create_project,
    },
    "notion_update_project": {
        "spec": {
            "type": "function",
            "function": {
                "name": "notion_update_project",
                "description": "Update an existing project Notion page by page ID.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "page_id": {"type": "string", "description": "Notion page UUID"},
                        "name": {"type": "string"},
                        "status": {"type": "string"},
                        "deadline": {
                            "type": "string",
                            "description": "YYYY-MM-DD or empty to clear",
                        },
                        "priority": {"type": "string"},
                        "current_focus": {"type": "string"},
                        "next_action": {"type": "string"},
                    },
                    "required": ["page_id"],
                },
            },
        },
        "handler": _update_project,
    },
    "notion_search_tasks": {
        "spec": {
            "type": "function",
            "function": {
                "name": "notion_search_tasks",
                "description": (
                    "Search your Notion tasks database: name, status, due, optional project link. "
                    "Filter by status optionally."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "status": {"type": "string", "description": "Optional status filter"},
                        "limit": {"type": "integer"},
                    },
                },
            },
        },
        "handler": _search_tasks,
    },
    "notion_create_task": {
        "spec": {
            "type": "function",
            "function": {
                "name": "notion_create_task",
                "description": "Create a task in the Tasks database. Optionally link to a project page ID.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "title": {"type": "string"},
                        "due": {"type": "string", "description": "YYYY-MM-DD"},
                        "status": {"type": "string"},
                        "project_page_id": {
                            "type": "string",
                            "description": "Parent project Notion page UUID",
                        },
                    },
                    "required": ["title"],
                },
            },
        },
        "handler": _create_task,
    },
    "notion_update_task": {
        "spec": {
            "type": "function",
            "function": {
                "name": "notion_update_task",
                "description": "Update a task page by ID.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "page_id": {"type": "string"},
                        "title": {"type": "string"},
                        "due": {"type": "string"},
                        "status": {"type": "string"},
                        "project_page_id": {
                            "type": "string",
                            "description": "Set or replace project relation",
                        },
                    },
                    "required": ["page_id"],
                },
            },
        },
        "handler": _update_task,
    },
    "notion_describe_tasks_database": {
        "spec": {
            "type": "function",
            "function": {
                "name": "notion_describe_tasks_database",
                "description": (
                    "Return the tasks database schema: each property’s type and, for Status "
                    "(select or Notion status), every allowed option name. Use before create/update "
                    "when status values are unknown."
                ),
                "parameters": {"type": "object", "properties": {}},
            },
        },
        "handler": _describe_tasks_database,
    },
}
